lib/core/src/Cardano/Wallet/Api/Server.hs

```diff
@@ -786,15 +801,53 @@ listPools spl =
             apparentPerformance
 
 joinStakePool
-    :: forall n k.
+    :: forall ctx s t n k.
         ( PaymentAddress n k
+        , Buildable (ErrValidateSelection t)
+        , s ~ SeqState n k
         , k ~ ShelleyKey
+        , HardDerivation k
+        , ctx ~ ApiLayer s t k
         )
-    => ApiT PoolId
+    => ctx
+    -> StakePoolLayer IO
+    -> ApiT PoolId
     -> ApiT WalletId
     -> ApiWalletPassphrase
     -> Handler (ApiTransaction n)
-joinStakePool _ _ _ = throwError err501
+joinStakePool ctx spl (ApiT poolId) (ApiT wid) passwd = do
+    (_, walMeta, _) <- liftHandler $ withWorkerCtx ctx wid throwE $
+        \wrk -> W.readWallet wrk wid
+    when (walMeta ^. #delegation == W.Delegating poolId) $
+        liftHandler $ throwE (ErrCreateCertPoolAlreadyJoined poolId)
+
+    allPools <- listPools spl
```

Comment:
I'm not sure if this is wise.
1. `listStakePools` is not a reliable way to list the current `[PoolId]` because it has to be in sync with its chain-follower.*
2. This is using the api-`listPools`. So the error-messages from that endpoint can be thrown here I presume. I had never realised it was possible to nest api-calls like this though — cool! — but I don't think we want it?

Could we use `NetworkLayer.stakeDistribution` instead?

Reply:
oh, thanks for this remark. As this is not the only one PR in this task, I will address it in the signCert (or signDelegationCert) PR soon

Reply:
what about simply relying on `DB.readPoolProduction` instead to avoid unnecessary networks roundtrips?

Reply:
> what about simply relying on DB.readPoolProduction instead to avoid unnecessary networks roundtrips?

In order for the `DB` to contain any stake-distributions (would be `DB.readStakeDistribution`), it does at least need to _begin_ following the chain correctly. In contrast to the wallet, it needs to (I think it was)`fetchTip >> fetchStakeDistribution >> fetchTip` without the tip changing, which is a point of that would be unnecessary to bring in here.

Doing that would entangle the stake-pool & wallet state in a way I am worried about.

The network roundtrip should be _really_ insignificant here, anyway?